### DISCUSSION TITLE
Possible improvement for the library

### DISCUSSION THREAD

**Discussion by daniil226** [syssi/esphome-pipsolar#139] [Q&A] [answered]
Forgive me if I say some nonsense, but maybe I'll say something to the point :)
It would be great to make the following changes in the library

```
in file
 components/pipsolar/sensor/__init__.py
line 307
    CONF_TOTAL_GENERATED_ENERGY: sensor.sensor_schema(
        unit_of_measurement=UNIT_WATT,
        accuracy_decimals=0,
        device_class=DEVICE_CLASS_POWER,
```

the inverter gives a counter of generated energy in watt hours and not in watts
It is very convenient to add this counter to the energy panel, so it would be good to replace the class with energy and add
state_class must be total or total_increasing.

It might be better to replace it with the following lines (possibly with errors) without taking into account possible changes in another part of the library:

```
   CONF_TOTAL_GENERATED_ENERGY: sensor.sensor_schema(
        unit_of_measurement=UNIT_WATT_HOUR,
        accuracy_decimals=0,
        device_class=DEVICE_CLASS_ENERGY,
        state_class=TOTAL_INCREASING
```

**Comment by syssi** [comment]
Do you use one of the feature branches or do you use the official esphome component?

**Reply by syssi** [reply]
Okay, we are talking about the `pi18` branch. I've pushed a change!

**Reply by daniil226** [reply]
external_components_source: github://syssi/esphome-pipsolar@pi18

**Comment by daniil226** [comment]
Looks like I made a mistake somewhere. Sorry, it's probably best to undo the changes.
gives an error when trying to install an update:

                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
NameError: name 'STATE_CLASS_TOTAL_INCREASING' is not defined
Failed config

**Comment by daniil226** [comment]
INFO ESPHome 2024.7.2
INFO Reading configuration /config/esphome/invertor.yaml...
INFO Updating https://github.com/syssi/esphome-pipsolar.git@pi18
ERROR Unable to load component pipsolar.sensor:
Traceback (most recent call last):
  File "/esphome/esphome/loader.py", line 176, in _lookup_module
    module = importlib.import_module(f"esphome.components.{domain}")
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1206, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1178, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1149, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/data/external_components/f2cdeb5c/components/pipsolar/sensor/__init__.py", line 316, in <module>
    state_class=STATE_CLASS_TOTAL_INCREASING,
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
NameError: name 'STATE_CLASS_TOTAL_INCREASING' is not defined
Failed config

sensor.pipsolar: [source /config/esphome/invertor.yaml:60]
  
  Platform not found: 'sensor.pipsolar'.
  platform: pipsolar
  pipsolar_id: inverter0
  grid_rating_voltage: 
    name: pipsolar grid_rating_voltage
  grid_rating_current: 
    name: pipsolar grid_rating_current
  ac_output_rating_voltage: 
    name: pipsolar ac_output_rating_voltage
  ac_output_rating_frequency: 
    name: pipsolar ac_output_rating_frequency
  ac_output_rating_current: 
    name: pipsolar ac_output_rating_current

**Reply by syssi** [reply]
Please give it another try. I've fixed all missing imports.

**Comment by daniil226** [comment]
everything works fine, problem fixed. The sensor did not appear in the energy panel although it has the necessary attributes.
![ok](https://github.com/user-attachments/assets/3eeb2e53-a4c3-42b1-8a25-ee31b1888860)

**Reply by syssi** [reply]
It looks like the sensor is unavailable and doesn't have a proper state/measurement?

**Reply by daniil226** [reply]
At the time of the screenshot it was updating, now everything is ok. I managed to add a utility meter to the integration and to the energy panel. Thanks a lot .

**Reply by syssi** [reply, marked as answer]
Please check the entity at the developer tools at `/developer-tools/statistics`. Does the sensor require a fix first?

**Reply by daniil226** [reply]
Yes, I updated and everything worked without integrations. huge gratitude to you.

**Reply by daniil226** [reply]
if I want to help the project where can I donate money?

**Reply by syssi** [reply]
Thanks in advance!

https://paypal.me/syssi84
https://buymeacoffee.com/syssi

**Reply by daniil226** [reply]
I sent a little through the coffee service, sorry it’s not enough. We're going through a difficult time right now. Thanks from Ukraine

**Reply by syssi** [reply]
Мої думки з вами (Moji dumky z vamy)

**Comment by daniil226** [comment]
Can I please have one last question:
there is a branch of your library for the PI18_LV5048_Hybrid_V2 inverter 

it probably uses the encoding of requests to the inverter that I need.
When I try to connect this branch to Yaml external_components_source: github://preussal/esphome-pipsolar@PI18_LV5048_Hybrid_V2 # Stable

When compiling it gives an error. screenshot below. and when connecting your branch github://syssi/esphome-pipsolar@pi18
everything works fine but does not receive dynamic data from the inverter 
Tell me how to correctly connect the required branch?
![error](https://github.com/user-attachments/assets/fb350db4-6e61-4f58-ac44-ae6e81df3caf)

**Reply by syssi** [reply]
Please try a clean build: `esphome clean yourconfig.yaml` or press the `Clean build files` button at the ESPHome dashboard + compile the code another time. Does it work?

**Reply by daniil226** [reply]
unfortunately it didn't help

**Reply by syssi** [reply]
The contribution isn't ready probably. I will have a look but it will take some time.

**Reply by daniil226** [reply]
Of course, no rush, only when there is absolutely nothing to do. Thanks a lot
